### ISSUE TITLE
Javassist v3.19.0-GA 19 Incompatibility

### ISSUE THREAD

**Issue by dharrigan** [brettwooldridge/HikariCP#227]
Konichiwa Brett! Hope you are well.

Java 8u25

I've come across an issue using the just released Javassist v3.19.0-GA. It doesn't happen with v3.18.2-GA.

```
java.lang.VerifyError: Illegal type at constant pool entry 392 in class com.zaxxer.hikari.proxy.PreparedStatementJavassistProxy
Exception Details:
  Location:
    com/zaxxer/hikari/proxy/PreparedStatementJavassistProxy.getLargeUpdateCount()J @1: invokespecial
  Reason:
    Constant pool index 392 is invalid
  Bytecode:
    0x0000000: 2ab7 0188 ad4c 2a2b b600 1bbf          
  Exception Handler Table:
    bci [0, 5] => handler: 5
  Stackmap Table:
    same_locals_1_stack_item_frame(@5,Object[#17])

	at com.zaxxer.hikari.proxy.ProxyFactory.getProxyPreparedStatement(ProxyFactory.java) ~[HikariCP-2.3.0-rc4.jar:na]
	at com.zaxxer.hikari.proxy.ConnectionProxy.prepareStatement(ConnectionProxy.java:263) ~[HikariCP-2.3.0-rc4.jar:na]
	at com.zaxxer.hikari.proxy.ConnectionJavassistProxy.prepareStatement(ConnectionJavassistProxy.java) ~[HikariCP-2.3.0-rc4.jar:na]
	at org.flywaydb.core.internal.dbsupport.JdbcTemplate.queryForString(JdbcTemplate.java:181) ~[flyway-core-3.1.jar:na]
	at org.flywaydb.core.internal.dbsupport.postgresql.PostgreSQLDbSupport.doGetCurrentSchema(PostgreSQLDbSupport.java:55) ~[flyway-core-3.1.jar:na]
	at org.flywaydb.core.internal.dbsupport.DbSupport.getCurrentSchema(DbSupport.java:75) ~[flyway-core-3.1.jar:na]
	at org.flywaydb.core.Flyway.execute(Flyway.java:1295) ~[flyway-core-3.1.jar:na]
	at org.flywaydb.core.Flyway.migrate(Flyway.java:919) ~[flyway-core-3.1.jar:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_25]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_25]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_25]
	at java.lang.reflect.Method.invoke(Method.java:483) ~[na:1.8.0_25]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeCustomInitMethod(AbstractAutowireCapableBeanFactory.java:1694) ~[spring-beans-4.1.4.RELEASE.jar:4.1.4.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1633) ~[spring-beans-4.1.4.RELEASE.jar:4.1.4.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1562) ~[spring-beans-4.1.4.RELEASE.jar:4.1.4.RELEASE]
	... 55 common frames omitted
```

**Comment by brettwooldridge** [comment]
This should be fixed in the 2.3.0-rc4 snapshot, can you check it?

**Comment by dharrigan** [comment]
Hi Brett,

Something strange is going on :-).

On my local development box, with Javassist 3.19.0-GA, and HikariCP RC4, my system boots fine.

When I deploy the same WAR into an EC2 instance running on Amazon, I get the exception as listed above. I can confirm that in the lib directory, both RC4 and 3.19.0-GA are present, and no rogue RC3 (or prior) or Javassist 3.18.2-GA (or prior) exists.

This is the result of java -version on the EC2 instance:

$ java -version
java version "1.8.0_25"
Java(TM) SE Runtime Environment (build 1.8.0_25-b17)
Java HotSpot(TM) 64-Bit Server VM (build 25.25-b02, mixed mode)

and here's what I have locally:

$ java -version 
java version "1.8.0_25"
Java(TM) SE Runtime Environment (build 1.8.0_25-b17)
Java HotSpot(TM) 64-Bit Server VM (build 25.25-b02, mixed mode)

Any suggestions?

-=david=-

**Comment by brettwooldridge** [comment]
If you delete the 3.19.0-GA jar completely, do you get a similar error?  Maybe another library is embedding (shadowing) the Javassist classes...

**Comment by dharrigan** [comment]
Hi Brett,

It's not that I'm afraid. Firstly, I did a grep to determine if any of my 3rd party jars had repackaged javassist. Here are the results:

```
$ grep -ri "javassist" *.jar
Binary file hibernate-core-4.3.8.Final.jar matches
Binary file HikariCP-2.3.0-rc4.jar matches
Binary file javassist-3.19.0-GA.jar matches
Binary file orika-core-1.4.5.jar matches
```

Then, I deployed a version of the WAR containing 3.19.0-GA and confirmed the error above. I then removed javassist-3.19.0-GA from the lib directory and restarted the application server, and I got this

```
java.lang.ClassNotFoundException: javassist.NotFoundException
	at java.net.URLClassLoader$1.run(URLClassLoader.java:372) ~[na:1.8.0_25]
	at java.net.URLClassLoader$1.run(URLClassLoader.java:361) ~[na:1.8.0_25]
	at java.security.AccessController.doPrivileged(Native Method) [na:1.8.0_25]
	at java.net.URLClassLoader.findClass(URLClassLoader.java:360) ~[na:1.8.0_25]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[na:1.8.0_25]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[na:1.8.0_25]
	at org.eclipse.jetty.webapp.WebAppClassLoader.loadClass(WebAppClassLoader.java:450) ~[na:na]
	at org.eclipse.jetty.webapp.WebAppClassLoader.loadClass(WebAppClassLoader.java:403) ~[na:na]
	... 82 common frames omitted
Wrapped by: java.lang.NoClassDefFoundError: javassist/NotFoundException
	at com.zaxxer.hikari.HikariConfig.<clinit>(HikariConfig.java:32) ~[HikariCP-2.3.0-rc4.jar:na]
```

Indicating that the javassist libraries have not been found at all.

-=david=-

**Comment by dharrigan** [comment]
Hi (again!)

I've managed to reproduce the issue locally on my machine.

If I add 

```-noverify```

as a JVM parameter, I *do not* get the javassist error, otherwise, removing the -noverify, causes it to bomb out.

I don't get the problem using javassist 3.18.2-GA with *no* -noverify.

In summary:

Javassist 3.19.0-GA without -noverify causes HikariCP to bomb out.
Javassist 3.18.2-GA without -noverify does not cause HikariCP to bomb out.

Hope this helps!

-=david=-

**Comment by brettwooldridge** [comment]
I am curious if this is only occurring in the context of Spring?  For example, can you build HikariCP 2.3.0 branch locally using ``mvn package``?  If so, the unit tests are successfully using Javassist 3.19.0-GA without the ``-noverify`` flag...

**Comment by dharrigan** [comment]
Hi

Cloned repo, mvn package worked.

Changed to javassist 3.19.0-GA in the pom:

```
[main] INFO com.zaxxer.hikari.HikariDataSource - HikariCP pool HikariPool-15 is starting.
[main] INFO HikariPool - HikariCP pool HikariPool-15 is shutting down.
Tests run: 10, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 8.68 sec <<< FAILURE! - in com.zaxxer.hikari.TestConnections
testBackfill(com.zaxxer.hikari.TestConnections)  Time elapsed: 0.26 sec  <<< ERROR!
java.lang.VerifyError: Illegal type at constant pool entry 392 in class com.zaxxer.hikari.proxy.PreparedStatementJavassistProxy
Exception Details:
  Location:
    com/zaxxer/hikari/proxy/PreparedStatementJavassistProxy.getLargeUpdateCount()J @5: invokespecial
  Reason:
    Constant pool index 392 is invalid
  Bytecode:
    0x0000000: b801 ad4c 2ab7 0188 2b11 00c6 0454 ad4d
    0x0000010: 2a2c b600 1b2b 1100 c704 54bf
  Exception Handler Table:
    bci [4, 15] => handler: 15
  Stackmap Table:
    full_frame(@15,{Object[#2],Object[#431]},{Object[#17]})

	at com.zaxxer.hikari.proxy.ProxyFactory.getProxyPreparedStatement(ProxyFactory.java)
	at com.zaxxer.hikari.proxy.ConnectionProxy.prepareStatement(ConnectionProxy.java:256)
	at com.zaxxer.hikari.proxy.ConnectionJavassistProxy.prepareStatement(ConnectionJavassistProxy.java)
	at com.zaxxer.hikari.TestConnections.testBackfill(TestConnections.java:258)

testCreate(com.zaxxer.hikari.TestConnections)  Time elapsed: 0.003 sec  <<< ERROR!
java.lang.VerifyError: Illegal type at constant pool entry 392 in class com.zaxxer.hikari.proxy.PreparedStatementJavassistProxy
Exception Details:
  Location:
    com/zaxxer/hikari/proxy/PreparedStatementJavassistProxy.getLargeUpdateCount()J @5: invokespecial
  Reason:
....
....
```

Hope this helps.

-=david=-

**Comment by dharrigan** [comment]
Hi,

On the 2.3.0 branch it works (I see in your 2.3.0 pom.xml that you are referencing javassist 3.19.0-GA).

I accidently built with the dev branch, which threw up lots of the errors (when I changed the pom javassist to 3.19.0-GA), as shown above.

Could it be that the 2.3.0-RC4 is wrong somehow? 

-=david=-

**Comment by brettwooldridge** [comment]
Yes, that is it.  2.3.0-rc4 was built on Jan. 7, and the pull request with the fix for Javassist 3.19.0-GA was integrated a day later, on Jan. 8.  The fix will therefore be in the 2.3.0 release.

**Comment by dharrigan** [comment]
That's great to hear!

Are you still waiting on the other issue feedback before the 2.3.0 release (or perhaps thinking about a 2.3.0-rc5 :-) :-)) release :-)

-=david=-

**Comment by UVST** [comment]
I'm using IntelliJ IDEA 14, Spring-Boot 1.2.1, HikariCP 2.3.2, Javassist 3.19.0-GA  with the JRebel 6 plugin!

When I start the application with the JRepel plugin, the same exception is thrown as mentioned above, the only workaround is to add the option -noverify. Without the usage of JRebel everything runs perfect. 

This only as a hint! 
so long...

**Comment by brettwooldridge** [comment]
@UVST what version of JRebel 6?  They have fixed several Javassist issues since the 6.0.0 release:

https://zeroturnaround.com/software/jrebel/download/changelog/6-x/

**Comment by UVST** [comment]
@brettwooldridge 

I'm using JRebel Legacy Agent 6.0.3 (201501261447)

**Comment by oak-tree** [comment]
having the same issue here, using JRebel 6.1.2 (201504061000) 



    <spring.core.version>4.1.6.RELEASE</spring.core.version>
    <spring.security.core.version>4.0.1.RELEASE</spring.security.core.version>
    <spring.data.jpa.version>1.8.0.RELEASE</spring.data.jpa.version>
    <spring.test.version>4.0.6.RELEASE</spring.test.version>
    <cglib.version>3.1</cglib.version>
    <org.javassist.version>3.19.0-GA</org.javassist.version>
    <aspectj.version>1.8.1</aspectj.version>
    <joda.time.version>2.4</joda.time.version>
   
   <mysql.connector.version>5.1.35</mysql.connector.version>
<HikariCP.version>2.3.7</HikariCP.version>

**Comment by oak-tree** [comment]
Caused by: java.lang.VerifyError: Illegal type at constant pool entry 172 in class com.zaxxer.hikari.proxy.StatementJavassistProxy
Exception Details:
  Location:
    com/zaxxer/hikari/proxy/StatementJavassistProxy.getLargeMaxRows()J @1: invokespecial
  Reason:
    Constant pool index 172 is invalid
  Bytecode:
    0x0000000: 2ab7 00ac ad4c 2a2b b600 17bf          
  Exception Handler Table:
    bci [0, 5] => handler: 5
  Stackmap Table:
    same_locals_1_stack_item_frame(@5,Object[#15])

	at com.zaxxer.hikari.proxy.ProxyFactory.getProxyStatement(ProxyFactory.java)
	at com.zaxxer.hikari.proxy.ConnectionProxy.createStatement(ConnectionProxy.java:226)

**Comment by rodgarcialima** [comment]
Version 2.3.7 still has the bug. I have to do this to work:
```xml
<!-- Connection Pool -->
<dependency>
	<groupId>com.zaxxer</groupId>
	<artifactId>HikariCP</artifactId>
	<version>2.3.7</version>
	<exclusions>
		<!-- Bug do HikariCP -->
		<exclusion>
			<groupId>org.javassist</groupId>
			<artifactId>javassist</artifactId>
		</exclusion>
	</exclusions>
</dependency>

<dependency>
	<groupId>org.javassist</groupId>
	<artifactId>javassist</artifactId>
	<version>3.18.2-GA</version>
</dependency>
```

Great projet BTW. Cheers from Brasil.

**Comment by henri5** [comment]
This should be fixed in JRebel 6.2.1 release (see [here](https://zeroturnaround.com/software/jrebel/download/changelog/6-x/), search for HikariCP)

**Comment by brettwooldridge** [comment]
Additionally, HikariCP 2.4.0 no longer has a runtime dependency on javassist.
